content/src/debug.js

```diff
@@ -201,38 +360,199 @@ function reportSeeeval(method, debugId, length, coordId, elem, args){
-      traceLine(location);
+
+  if (!programChanged) {
+    var record = debugRecordsByDebugId[debugId];
+    if (record) {
+      if (!record.seeeval) {
+        record.method = method;
+        record.args = args;
+        var currentRecord = debugRecordsByDebugId[currentRecordID];
+        var currentIndex = currentRecord.eventIndex;
+        var currentLocation = traceEvents[currentIndex].location;
+        var currentLine = currentLocation.first_line;
+        var prevLine = 0;
+
+        stuckComplexity.moves += 1;
+        var index = record.eventIndex;
+        var line = traceEvents[index].location.first_line;
+        var appear_location = traceEvents[index].location;
+        var tracedIndex = -1; 
+
+        // trace lines that are not animation.
+        while (currentRecordID < debugId) {
+          if (prevIndex != -1) {
+            var prevLocation = traceEvents[prevIndex].location;
+            prevLine = prevLocation.first_line;
+          } else {
+            var prevLocation = null;
+            prevLine = -1;
+          }
+
+          if (tracedIndex!= -1) {
+            untraceLine(tracedIndex);
+            tracedIndex = -1;
+          }
+
+          // decide if it is necessary to draw add and draw any arrows.
+          if (currentLine < prevLine && currentIndex == prevIndex + 1) {
+
+            if (arrows[prevIndex] != null) {
+              arrows[prevIndex]['after'] =  {first: currentLocation, second: prevLocation};
+            } else {
+              arrows[prevIndex] = {before: null, after: {first: currentLocation, second: prevLocation}};
+            }
+            if (arrows[currentIndex] != null) {
+              arrows[currentIndex]['before'] =  {first: currentLocation, second: prevLocation};
+            } else{
+              arrows[currentIndex] = {before: {first: currentLocation, second: prevLocation}, after : null};
+            }
+            //Draw an arrow if the debugger is on
+            if (debugMode) {
+              view.arrow(view.paneid('left'), arrows, currentIndex, false);
+            }
+            if (!arrows[currentRecordID]){
+              debugRecordsByLineNo[currentLine] = currentRecordID;
+            }
+          }
+          traceLine(currentIndex);
+          tracedIndex = currentIndex;
+          prevLine = currentLine;
+          prevIndex = debugRecordsByDebugId[currentRecordID].eventIndex;
+          currentRecordID += 1;
+          currentRecord = debugRecordsByDebugId[currentRecordID];
+          currentIndex = currentRecord.eventIndex
+          currentLine = traceEvents[currentIndex].location.first_line;
+          currentLocation = traceEvents[currentIndex].location;
+        }
+        if (tracedIndex != -1) {
+          untraceLine(tracedIndex);
+          tracedIndex = -1;
+        }
+        if (line < prevLine && index == prevIndex + 1){
+
+          prevLocation = traceEvents[prevIndex].location;
+            
+          if (arrows[prevIndex] != null){
```

Comment:
This looks suspiciously like code below in end_program(). Maybe make it into a function to be called in both places?